### PR TITLE
feat: OAuth security integration for partner applications

### PR DESCRIPTION
**Pull request by gouline**
Implementing OAuth security integration, largely based on the existing SCIM security integration that's quite similar. 

This one only covers the **partner applications** (i.e. Tableau Desktop, Tableau Server and Looker), however extending it to support custom OAuth clients would just involve adding [required](https://docs.snowflake.com/en/sql-reference/sql/create-security-integration.html#snowflake-oauth-custom-client-parameters) and [optional](https://docs.snowflake.com/en/sql-reference/sql/create-security-integration.html#snowflake-oauth-custom-client-parameters-optional) parameters.

## Test Plan

* [x] acceptance tests

## References

* [Documentation](https://docs.snowflake.com/en/user-guide/oauth-partner.html)
* [CREATE](https://docs.snowflake.com/en/sql-reference/sql/create-security-integration.html)
* [ALTER](https://docs.snowflake.com/en/sql-reference/sql/alter-security-integration.html)
* [SHOW](https://docs.snowflake.com/en/sql-reference/sql/show-integrations.html)
* [DESCRIBE](https://docs.snowflake.com/en/sql-reference/sql/desc-integration.html)

### CODE REVIEW

File: pkg/resources/oauth_integration.go [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+	},
+	"oauth_issue_refresh_tokens": {
+		Type:        schema.TypeBool,
+		Optional:    true,
+		Description: "Specifies whether to allow the client to exchange a refresh token for an access token when the current access token has expired.",
+	},
+	"oauth_refresh_token_validity": {
+		Type:        schema.TypeInt,
+		Optional:    true,
+		Description: "Specifies how long refresh tokens should be valid (in seconds). OAUTH_ISSUE_REFRESH_TOKENS must be set to TRUE.",
```
Comment (edulop91) [inline pkg/resources/oauth_integration.go:38]: do we need `oauth_issue_refresh_tokens` or would the presence/absence of `oauth_refresh_token_validity` give us enough information?

File: pkg/resources/oauth_integration.go [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+	"oauth_use_secondary_roles": {
+		Type:        schema.TypeString,
+		Optional:    true,
+		Description: "Specifies whether default secondary roles set in the user properties are activated by default in the session being opened.",
+		ValidateFunc: validation.StringInSlice([]string{
+			"IMPLICIT", "NONE",
+		}, false),
+	},
+	"blocked_roles_list": {
+		Type:        schema.TypeList,
```
Comment (edulop91) [inline pkg/resources/oauth_integration.go:49]:
```suggestion
		Type:        schema.TypeSet,
```
order shouldn't matter

File: pkg/resources/oauth_integration.go (hunk 1 above)
Comment (edulop91) [reply pkg/resources/oauth_integration.go:38]: I see that keeping both would more closely match Snowflake's api, so we might want to keep as is

**Review by edulop91 — COMMENTED** [review]
just a couple of quick comments, thank you!

File: pkg/resources/oauth_integration.go (hunk 1 above)
Comment (gouline) [reply pkg/resources/oauth_integration.go:38]: Looks like `oauth_issue_refresh_tokens` is true by default, so you need it here if you want to set it to false. However, `oauth_refresh_token_validity` is optional too, it defaults to various values depending on the client (10 hours for Tableau Desktop, 90 days for Tableau Server, etc) - there will be cases where people omit both or specify only one.

**Comment by edulop91** [comment]
> Integration tests failure for [d10af87](https://github.com/chanzuckerberg/terraform-provider-snowflake/actions/runs/1495613910)

Looks like there might be a panic here: https://github.com/chanzuckerberg/terraform-provider-snowflake/runs/4301715646?check_suite_focus=true#step:9:1050 (not sure if this log is hidden behind GitHub access controls)

**Comment by gouline** [comment]
I can't get acceptance tests to run locally, so hopefully this fixes it. Found a similar instance in network policies and copied it:

```go
expandStringList(d.Get("blocked_roles_list").(*schema.Set).List())
```

File: pkg/resources/oauth_integration.go [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+		stmt.SetBool(`OAUTH_ISSUE_REFRESH_TOKENS`, d.Get("network_policy").(bool))
+	}
+	if _, ok := d.GetOk("oauth_refresh_token_validity"); ok {
+		stmt.SetInt(`OAUTH_REFRESH_TOKEN_VALIDITY`, d.Get("oauth_refresh_token_validity").(int))
+	}
+	if _, ok := d.GetOk("oauth_use_secondary_roles"); ok {
+		stmt.SetString(`OAUTH_USE_SECONDARY_ROLES`, d.Get("oauth_use_secondary_roles").(string))
+	}
+	if _, ok := d.GetOk("blocked_roles_list"); ok {
+		stmt.SetStringList(`BLOCKED_ROLES_LIST`, expandStringList(d.Get("blocked_roles_list").([]interface{})))
```
[the remark this answers is not in the document]
Comment (gouline) [reply pkg/resources/oauth_integration.go]: I can't get acceptance tests to run locally, so hopefully this fixes it. Found a similar instance in network policies and copied it:

```go
expandStringList(d.Get("blocked_roles_list").(*schema.Set).List())
```

**Comment by gouline** [comment]
CI step `integration-trusted` failed but I have no access to it, so cannot see what is causing it.

**Comment by alldoami** [comment]
```
=== CONT  TestAcc_OAuthIntegration
panic: interface conversion: interface {} is nil, not bool

goroutine 60401 [running]:
github.com/chanzuckerberg/terraform-provider-snowflake/pkg/resources.CreateOAuthIntegration(0x0, {0x14b1940, 0xc002a0c680})
	/home/runner/work/terraform-provider-snowflake/terraform-provider-snowflake/pkg/resources/oauth_integration.go:99 +0x7c5
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0x1330a20, {0x17f2748, 0xc000f6ce80}, 0x2, {0x14b1940, 0xc002a0c680})
	/home/runner/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.8.0/helper/schema/resource.go:318 +0x178
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc00073cb60, {0x17f2748, 0xc000f6ce80}, 0xc00053c270, 0xc003091580, {0x14b1940, 0xc002a0c680})
	/home/runner/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.8.0/helper/schema/resource.go:456 +0x871
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc003b0c1b0, {0x17f2748, 0xc000f6ce80}, 0xc003e37900)
	/home/runner/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.8.0/helper/schema/grpc_provider.go:977 +0xd8a
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc0026ff4e0, {0x17f27f0, 0xc001833b60}, 0x4423e0)
	/home/runner/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.4.0/tfprotov5/tf5server/server.go:332 +0x6c
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x14722e0, 0xc0026ff4e0}, {0x17f27f0, 0xc001833b60}, 0xc002634e40, 0x0)
	/home/runner/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.4.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:380 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc00217e540, {0x180adc8, 0xc0025a7b00}, 0xc00299ca20, 0xc0036934a0, 0x2152ee0, 0x0)
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.41.0/server.go:1279 +0xccf
google.golang.org/grpc.(*Server).handleStream(0xc00217e540, {0x180adc8, 0xc0025a7b00}, 0xc00299ca20, 0x0)
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.41.0/server.go:1608 +0xa2a
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.41.0/server.go:923 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.41.0/server.go:921 +0x294
FAIL	github.com/chanzuckerberg/terraform-provider-snowflake/pkg/resources	356.529s
```
When you click on the link in this comment, what does it say? https://github.com/chanzuckerberg/terraform-provider-snowflake/pull/763#issuecomment-978467694

**Comment by gouline** [comment]
> When you click on the link in this comment, what does it say? [#763 (comment)](https://github.com/chanzuckerberg/terraform-provider-snowflake/pull/763#issuecomment-978467694)

Sorry, this works. I was previously clicking on 'integration-trusted' in the footer (force of habit).

**Comment by gouline** [comment]
Found the problem! Silly copy-paste typo 🤦

**Comment by gouline** [comment]
Hmm, are these all meant to be treated as strings and parsed to booleans/integers? https://github.com/gouline/terraform-provider-snowflake/blob/161f112/pkg/resources/oauth_integration.go#L186-L201

I assumed from the `ENABLED` parameter fed as a boolean in all unit tests, they should be treated as booleans/integers, but `ENABLED` is always ignored in the `switch` statement so it might just be copy-pasted mistake that nobody picked up?

```go
descRows := sqlmock.NewRows([]string{
		"property", "property_type", "property_value", "property_default",
	}).AddRow("ENABLED", "Boolean", true, false)
```

Latest commit treats them as strings, so hopefully that works 🤷

**Comment by gouline** [comment]
Previous error looks fixed now, latest commit should fix the new error.

**Comment by gouline** [comment]
Should be fixed again.

**Comment by gouline** [comment]
This one was interesting: `blocked_roles_list` should only include roles not already implicitly enforced (ACCOUNTADMIN, ORGADMIN, SECURITYADMIN) on create, so I had to filter them out from the read. Should work now.

**Review by alldoami — APPROVED** [review]

**Pull request merged**